### PR TITLE
version 0.07 proposal

### PR DESCRIPTION
**Pull request by lawwman**
## Change on export var name
`functionWithBackpack ` -> `fnWithBackpack`

## Changes on item version
we are no longer using `version`. It is basically `emptyBackpack`. Instead, we will bring forward the `expiry` implementation.

## throwItem()
users now have the freedom to throw a single item instead of emptying the whole backpack.

## Change on `CarryBackpackParams`

previous:
```typescript
export interface CarryBackpackParams<
  T extends (...args: any[]) => Promise<any>
> {
  fn: T;
  getLatestItemVersion: () => string;
  initialVersion?: string;
  makeNoise?: boolean;
}
```

after:
```typescript
export interface CarryBackpackParams<
  T extends (...args: any[]) => Promise<any>
> {
  fn: T;
  expiry?: { ttl: number } // in seconds
  makeNoise?: boolean;
}
```

## unit test refactor
an attempt to nicely test multiple different scenarios without repeating logic.

example of scenarios:
- simple function
- function with multiple args
- slow function
- failing function
- backpack with no options
- backpack with versioning

## readme
updated readme with basic instructions

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,41 +1,44 @@
...
-  version: string;
+  lastRetrieved: number;
 }
 
-interface FunctionWithBackpack<T extends (...args: any) => Promise<any>> {
-  functionWithBackpack: T;
+interface FnWithBackpack<T extends (...args: any) => Promise<any>> {
+  fnWithBackpack: T;
   emptyBackpack: () => void;
+  throwItem: (...args: any[]) => void;
```
Comment (phanshiyu) [inline src/index.ts]:
```
  throwItem: (...args: Parameters<T>) => void;
```

File: src/index.ts [hunk 2]
```diff
@@ -54,5 +57,10 @@ export function carryBackpack<T extends (...args: any[]) => Promise<any>>(
     backpack.clear();
   }
 
-  return { functionWithBackpack, emptyBackpack };
+  function throwItem(...args: any[]) {
```
Comment (phanshiyu) [inline src/index.ts]:
```
  function throwItem(...args: Parameters<T>) {
```

File: src/index.ts [hunk 3]
```diff
@@ -1,41 +1,44 @@
...
+  throwItem: (...args: any[]) => void;
 }
 
 export interface CarryBackpackParams<
   T extends (...args: any[]) => Promise<any>
 > {
   fn: T;
-  getLatestItemVersion: () => string;
-  initialVersion?: string;
+  expiry?: { ttl: number } // in seconds
```
Comment (phanshiyu) [inline src/index.ts]: i think ttl can either be more explicit, `ttlInSec` or we do
```
  expiry?: {
    /** Time to live in seconds **/
    ttl: number;
  };
```
which shows up like this
Screenshot 2022-08-30 at 9 20 27 AM

**Comment by phanshiyu** [comment]
should we log out when expiry happens?

**Comment by lawwman** [comment]
> should we log out when expiry happens?

updated to log on expiry 👍

**Review by phanshiyu — APPROVED** [review]
SICK

**Pull request merged**